### ISSUE TITLE
IO thread makes OpenGL ES calls with no EGL context due to SkiaUnrefQueue/GrDirectContext shutdown

### ISSUE THREAD

**Issue by jason-simmons** [flutter/flutter#103517]
On a Moto G4 I'm sometimes seeing "call to OpenGL ES API with no current context" warnings in the log after pressing the back button in a Flutter app.

`Shell` destruction calls `platform_view->ReleaseResourceContext` on the IO thread, which clears the thread's EGL context in preparation for destroying the context.  However, with https://github.com/flutter/engine/pull/32106 the `SkiaUnrefQueue` destructor may now post a deferred task that deletes Skia objects and the `GrDirectContext`.  This task would still need the EGL context.

After decoding completes, the image decoder task on the `ConcurrentTaskRunner` posts a task to the IO thread to upload the image to a texture.  This task checks whether the `IOManager` is alive.  If it is, then the IO thread's `GrDirectContext` and EGL context are still available and it is safe to proceed.

The IO thread then needs to post a task to the UI thread to deliver the texture to Dart.  This is not synchronized with engine shutdown, and it means that the `SkiaUnrefQueue` may still be alive even after shutdown has deleted the engine and the `IOManager`.

The shell needs some way to defer clearing of the IO thread's EGL context until after the `GrDirectContext` has been deleted.

Some options:
* wrap the `GrDirectContext` in an object that also manages the EGL context and have the `SkiaUnrefQueue` hold a reference to the wrapper
* create a thread-local object on the IO thread whose destructor deletes the EGL context on thread shutdown
* have the `Shell` destructor queue a task to the UI thread and wait for it after deleting the `IOManager`.  Then queue a task to the IO thread that clears the thread's EGL context.  By the time the second task runs all strong references to the `SkiaUnrefQueue` held by either IO or UI thread tasks should be dropped, and it should be safe to release the EGL context.

@dnfield @chinmaygarde

**Comment by dnfield** [comment]
The thread local object seems like it'd be pretty safe and simple, right?

**Comment by zanderso** [comment]
Could this be related to https://github.com/flutter/flutter/issues/103498 ?

**Comment by jason-simmons** [comment]
This issue involves OpenGL ES calls made when the `SkiaUnrefQueue` deletes Skia objects and the `GrDirectContext`.

The `glTexImage2D` calls in https://github.com/flutter/flutter/issues/103498 would happen when the image decoding tasks on the IO thread call `SkImage::MakeCrossContextFromPixmap`.

Those calls are gated by checks of a `WeakPtr<IOManager>`, so they should only be made when the `GrDirectContext` and IO thread EGL context are in a valid state.  Any IO thread tasks that run after `Shell` shutdown deletes the `IOManager` should just drop the decoded image buffer and exit.
